src/JBDistributor.sol

```diff
@@ -1,208 +1,297 @@
+                // TODO muldiv lib
+                uint256 _tokenAmount = _distributable * _tokenStake(_tokenId) / _totalStakeAmount;
 
                 // Claim the share for this token
-                tokenVesting[_tokenIds[_j]][_vestingReleaseRound][_token] = _tokenAmount;
+                vestingTokenAmountAtRoundOf[_tokenId][_vestingReleaseRound][_token] = _tokenAmount;
 
-                emit claimed(_tokenIds[_j], _token, _tokenAmount, _vestingReleaseRound);
+                emit Claimed(_tokenId, _token, _tokenAmount, _vestingReleaseRound);
 
                 unchecked {
-                    // Increment the amount of tokens that have been claimed
+                    // Increment the amount of tokens that have been claimed and are now vesting.
                     _totalVestingAmount += _tokenAmount;
 
                     ++_j;
                 }
             }
 
             unchecked {
-                // Update the global claimable amount to reflect this claim
-                tokenVestingAmount[_token] += _totalVestingAmount;
+                // Store the updated total claimed amount now vesting.
+                vestingAmountOf[_token] += _totalVestingAmount;
 
                 ++_i;
             }
         }
     }
 
-    /**
-     * Collect vested tokens
-     * @param _tokenIds the nft ids to claim for
-     * @param _tokens the tokens to claim
-     * @param _round the round in which the tokens were done vesting
-     */
-    function collectVestedRewards(uint256[] calldata _tokenIds, IERC20[] calldata _tokens, uint256 _round) public {
-        // Make sure the vesting is done
-        if (_round > currentRound()) {
-            revert NotVestedYet();
+    /// @notice Allows an address to collect vested tokens for various rounds.
+    /// @param _rounds The rounds to collect for.
+    function collectVestedRewards(CollectVestingRoundData[] calldata _rounds) external {
+        // TODO: We can optimize this call by batching transfers
+        // Collect for each specified round.
+        for (uint256 _i = 0; _i < _rounds.length;) {
+            collectVestedRewards(_rounds[_i].tokenIds, _rounds[_i].tokens, _rounds[_i].round, _rounds[_i].beneficiary);
+
+            unchecked {
+                ++_i;
+            }
         }
+    }
+    
+    //*********************************************************************//
+    // ----------------------- public transactions ----------------------- //
+    //*********************************************************************// 
 
-        for (uint256 _i; _i < _tokens.length;) {
+    /// @notice Collect vested tokens.
+    /// @param _tokenIds The IDs of the 721s to claim for.
+    /// @param _tokens The address of the tokens being claimed.
+    /// @param _round The round during which the tokens were done vesting.
+    function collectVestedRewards(uint256[] calldata _tokenIds, IERC20[] calldata _tokens, uint256 _round, address _beneficiary) public {
```

Comment:
these transfer can be batched too if some ids have the same token(s)

Reply:
yes plz.

Reply:
They should already be batched(?), it first loops over each token and then checks if any of the given NFT ids can claim. At the end it performs the transfer for the sum of tokens

Reply:
ahh. makes sense. maybe @drgorillamd spots more optimization opportunity?

Reply:
Same mistakes as the i==0 (as B mention in the comment), all good here